app/scripts/controllers/permissions/background-api.js

```diff
@@ -52,73 +138,122 @@ export function getPermissionBackgroundApiMethods(permissionController) {
+          updatedCaveatValue,
         );
       }
     },
 
     addPermittedChain: (origin, chainId) => addMoreChains(origin, [chainId]),
+
     addPermittedChains: (origin, chainIds) => addMoreChains(origin, chainIds),
 
     removePermittedChain: (origin, chainId) => {
-      const { value: existingChains } = permissionController.getCaveat(
-        origin,
-        PermissionNames.permittedChains,
-        CaveatTypes.restrictNetworkSwitching,
-      );
+      let caip25Caveat;
+      try {
+        caip25Caveat = permissionController.getCaveat(
+          origin,
+          Caip25EndowmentPermissionName,
+          Caip25CaveatType,
+        );
+      } catch (err) {
+        // noop
+      }
 
-      const remainingChains = existingChains.filter(
-        (existingChain) => existingChain !== chainId,
+      if (!caip25Caveat) {
+        throw new Error(
+          `Cannot remove permission for chainId "${chainId}": No permissions exist for origin "${origin}".`,
+        );
+      }
+
+      const existingEthChainIds = getPermittedEthChainIds(caip25Caveat.value);
+
+      const remainingChainIds = existingEthChainIds.filter(
+        (existingChainId) => existingChainId !== chainId,
       );
 
-      if (remainingChains.length === existingChains.length) {
+      if (remainingChainIds.length === existingEthChainIds.length) {
         return;
       }
 
-      if (remainingChains.length === 0) {
+      if (remainingChainIds.length === 0 && !isSnap(origin)) {
         permissionController.revokePermission(
           origin,
-          PermissionNames.permittedChains,
+          Caip25EndowmentPermissionName,
         );
       } else {
+        const updatedCaveatValue = setPermittedEthChainIds(
+          caip25Caveat.value,
+          remainingChainIds,
+        );
         permissionController.updateCaveat(
           origin,
-          PermissionNames.permittedChains,
-          CaveatTypes.restrictNetworkSwitching,
-          remainingChains,
+          Caip25EndowmentPermissionName,
+          Caip25CaveatType,
+          updatedCaveatValue,
         );
       }
     },
 
-    requestAccountsAndChainPermissionsWithId: async (origin) => {
+    requestAccountsAndChainPermissionsWithId: (origin) => {
       const id = nanoid();
-      permissionController.requestPermissions(
-        { origin },
-        {
-          [PermissionNames.eth_accounts]: {},
-          [PermissionNames.permittedChains]: {},
-        },
-        { id },
-      );
-      return id;
-    },
+      approvalController
+        .addAndShowApprovalRequest({
+          id,
+          origin,
+          requestData: {
+            metadata: {
+              id,
+              origin,
+            },
+            permissions: {
+              [RestrictedMethods.eth_accounts]: {},
+              [PermissionNames.permittedChains]: {},
+            },
+          },
+          type: MethodNames.RequestPermissions,
+        })
+        .then((legacyApproval) => {
```

Comment:
Not ideal to use `.then` here rather than `async/await`. This will result in worse stack traces when errors are thrown (errors thrown in `.then`/`.catch` callbacks don't have parent frames, but they do when `await`-ed)

I am guessing that this was done to ensure the caller of this function wasn't blocked on the confirmation being approved, but there are different ways of accomplishing that that don't result in degraded stack traces (e.g. move it to another async function and don't `await` it)

Reply:
Though.... more importantly, why are we bypassing `permissionController.requestPermissions` here?

Reply:
the new permission does not have a factory and must be formed manually via an approval then a grant as opposed to a single requestPermissions call

Reply:
moved the logic into another async fn here https://github.com/MetaMask/metamask-extension/pull/27847/commits/d4c8df21d70c80d36643ddadbd10cdf64a5bdf0f

Reply:
Ah I see. Yeah we'd need to transform this in the UI instead if we wanted to use `permissionController.requestPermissions`.

I see in our Slack DM that you've mapped out a path to doing this transform in the UI instead, but for now I'm not inclined to block this PR on that. Between the two, transforming it in the UI does seem like it would be easier for us to maintain.

If you want to keep it here, I'll add one more nit: We should include a comment that explains why we're inlining `requestPermissions` rather than calling it directly.

Reply:
Comment added here https://github.com/MetaMask/metamask-extension/pull/27847/commits/ee043eb1f54913fa320cd4672fa5ea92a3e0b897